zstd_image_transport/src/zlib_cpp.cpp

```diff
@@ -0,0 +1,167 @@
+  return data;
+}
+
+std::shared_ptr<DataBlock> ExpandDataList(const std::list<std::shared_ptr<DataBlock>> & data_list)
+{
+  std::size_t total_size = 0;
+  for (const std::shared_ptr<DataBlock> & this_data : data_list) {
+    total_size += this_data->size;
+  }
+  std::shared_ptr<DataBlock> out_data = AllocateData(total_size);
+  uint8_t * this_ptr = out_data->ptr;
+  for (const std::shared_ptr<DataBlock> & this_data : data_list) {
+    memcpy(this_ptr, this_data->ptr, this_data->size);
+    this_ptr += this_data->size;
+  }
+  return out_data;
+}
+
+Comp::Comp(Level level, bool zlib_header)
+: level_(level)
+{
+  memset(&zs_, 0, sizeof(zs_));
+  int windowBits = WINDOW_BITS;
+  if (zlib_header) {
+    // Configurate the compressor to write a simple zlib header and trailer
+    // around the compressed data instead of a zlib wrapper
+    windowBits += 16;
+  }
+  int ret = deflateInit2(
+    &zs_, static_cast<int>(level_), Z_DEFLATED, windowBits,
+    8, Z_DEFAULT_STRATEGY);
+  init_ok_ = ret == Z_OK;
+}
+
+Comp::~Comp() {deflateEnd(&zs_);}
+
+bool Comp::IsSucc() const
+{
+  return init_ok_;
+}
+
+std::list<std::shared_ptr<DataBlock>> Comp::Process(
+  const uint8_t * buffer, std::size_t size,
+  bool last_block)
+{
+  std::list<std::shared_ptr<DataBlock>> out_data_list;
+  // Prepare output buffer memory.
+  uint8_t out_buffer[MAX_CHUNK_SIZE];
+  zs_.next_in = reinterpret_cast<uint8_t *>(const_cast<uint8_t *>(buffer));
+  zs_.avail_in = static_cast<uInt>(size);
+  do {
+    // Reset output buffer position and size.
+    zs_.avail_out = MAX_CHUNK_SIZE;
+    zs_.next_out = out_buffer;
+    // Do compress.
+    deflate(&zs_, last_block ? Z_FINISH : Z_NO_FLUSH);
+    // Allocate output memory.
+    std::size_t out_size = MAX_CHUNK_SIZE - zs_.avail_out;
+    std::shared_ptr<DataBlock> out_data = AllocateData(out_size);
+    // Copy and add to output data list.
+    memcpy(out_data->ptr, out_buffer, out_size);
+    out_data_list.push_back(std::move(out_data));
+  } while (zs_.avail_out == 0);
+  // Done.
+  return out_data_list;
+}
+
+Decomp::Decomp()
+{
+  memset(&zs_, 0, sizeof(zs_));
+  // Enable zlib and zlib decoding with automatic header detection
+  int windowBits = WINDOW_BITS + 32;
+  int ret = inflateInit2(&zs_, windowBits);
+  init_ok_ = ret == Z_OK;
+}
+
+Decomp::~Decomp() {inflateEnd(&zs_);}
+
+std::list<std::shared_ptr<DataBlock>> Decomp::Process(
+  const std::shared_ptr<DataBlock> & compressed_data)
+{
+  std::list<std::shared_ptr<DataBlock>> out_data_list;
+  uint8_t out_buffer[MAX_CHUNK_SIZE];
+  // Incoming buffer.
+  zs_.avail_in = static_cast<uInt>(compressed_data->size);
+  zs_.next_in = compressed_data->ptr;
+  int ret;
+  do {
+    // Prepare outcoming buffer and size.
+    zs_.avail_out = MAX_CHUNK_SIZE;
+    zs_.next_out = out_buffer;
+    // Decompress data.
+    ret = inflate(&zs_, Z_NO_FLUSH);
+    switch (ret) {
+      case Z_NEED_DICT:
+        // Incoming data is invalid.
+        return std::move(out_data_list);
```

Comment:
``warning: moving a local object in a return statement prevents copy elision [-Wpessimizing-move]``
Let the compiler do return value optimization instead by not using move semantics.

```suggestion
        return out_data_list;
```